cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,180 @@
+	if err := config.OperateOnCIOperatorConfigDir(abs, func(cfg *api.ReleaseBuildConfiguration, metadata *config.Info) error {
+		promotedTags.Insert(tagsToPromote(cfg).List()...)
+		return nil
+	}); err != nil {
+		logrus.WithField("path", abs).Fatal("failed to operate on CI Operator's config directory")
+	}
+
+	kubeConfigs, _, err := util.LoadKubeConfigs(opts.kubeconfig, nil)
+	if err != nil {
+		logrus.Fatalf("could not load kube config from path %s: %w", opts.kubeconfig, err)
+	}
+	if len(kubeConfigs) != 1 {
+		logrus.Fatalf("found %d contexts in kube config %s: must be %s only", len(kubeConfigs), opts.kubeconfig, string(api.ClusterAPPCI))
+	}
+
+	client, err := ctrlruntimeclient.New(kubeConfigs[string(api.ClusterAPPCI)], ctrlruntimeclient.Options{})
+	if err != nil {
+		logrus.WithError(err).Fatalf("could not create client")
+	}
+
+	ctx := context.TODO()
+	buildConfigs := &buildv1.BuildConfigList{}
+	if err := client.List(ctx, buildConfigs); err != nil {
+		logrus.WithError(err).Fatalf("could not list buildConfigs")
+	}
+	buildConfigOutputTags := sets.NewString()
+	for _, bc := range buildConfigs.Items {
+		if bc.Spec.Output.To != nil {
+			buildConfigOutputTags.Insert(fmt.Sprintf("%s/%s", bc.Spec.Output.To.Namespace, bc.Spec.Output.To.Name))
+		}
+	}
+	tagsToCheck := sets.NewString()
+	var errs []error
+	imageStreamsWithPromotedTags := sets.NewString()
+	for _, promotedTag := range promotedTags.List() {
+		namespacedImageStreamAndTagName := strings.Split(promotedTag, ":")
+		if n := len(namespacedImageStreamAndTagName); n != 2 {
+			errs = append(errs, fmt.Errorf("when splitting isTagName %s by : expected two results, got %d", namespacedImageStreamAndTagName, n))
+			continue
+		}
+		namespacedImageStream := namespacedImageStreamAndTagName[0]
+
+		namespaceAndImageStreamName := strings.Split(namespacedImageStream, "/")
+		if n := len(namespaceAndImageStreamName); n != 2 {
+			errs = append(errs, fmt.Errorf("when splitting tag %s by / expected two results, got %d", namespacedImageStream, n))
+			continue
+		}
+		ns, isName := namespaceAndImageStreamName[0], namespaceAndImageStreamName[1]
+
+		if imageStreamsWithPromotedTags.Has(namespacedImageStream) {
+			continue
+		}
+		imageStreamsWithPromotedTags.Insert(namespacedImageStream)
+		imagestream := &imagev1.ImageStream{}
+		if err := client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: ns, Name: isName}, imagestream); err != nil {
+			errs = append(errs, fmt.Errorf("could not get image steam %s in namespace %s", isName, ns))
+			continue
+		}
+		for _, tag := range imagestream.Status.Tags {
+			tagsToCheck.Insert(fmt.Sprintf("%s/%s:%s", ns, isName, tag.Tag))
+		}
+	}
+
+	intersection := promotedTags.Intersection(buildConfigOutputTags)
+	if intersection.Len() > 0 {
+		errs = append(errs, fmt.Errorf("found build config overwrite promoted tags: %s", intersection.List()))
+	}
+
+	tagsToDelete := sets.NewString()
+	for _, tag := range tagsToCheck.List() {
+		if buildConfigOutputTags.Has(tag) {
```

Comment:
This does not seem like a feature we need.

Reply:
Removed.